flyway-core/src/main/java/org/flywaydb/core/internal/dbsupport/saphana/SapHanaSqlStatementBuilder.java

```diff
@@ -44,24 +41,56 @@ protected String cleanToken(String token) {
 
     @Override
     protected Delimiter changeDelimiterIfNecessary(String line, Delimiter delimiter) {
-        if (StringUtils.countOccurrencesOf(statementStart, " ") < 4) {
-            statementStart += line;
-            statementStart += " ";
+
+        // need only accumulate 16 characters of normalized statement start in order to determine if it is an 'interesting' statement
+        if (statementStartNormalized.length() < 16) {
+            final String effectiveLine = cutCommentsFromEnd(line);
+            statementStartNormalized += effectiveLine + " ";
+            statementStartNormalized = StringUtils.trimLeadingWhitespace(StringUtils.collapseWhitespace(statementStartNormalized));
         }
+        boolean insideStatementAllowingNestedBeginEndBlocks =
+                statementStartNormalized.startsWith("CREATE PROCEDURE")
+                || statementStartNormalized.startsWith("CREATE FUNCTION")
+                || statementStartNormalized.startsWith("CREATE TRIGGER");
 
-        if (statementStart.startsWith("CREATE TRIGGER")) {
+        if (insideStatementAllowingNestedBeginEndBlocks) {
             if (line.startsWith("BEGIN")) {
-                insideBeginEndBlock = true;
+                beginEndNestedDepth++;
             }
 
             if (line.endsWith("END;")) {
-                insideBeginEndBlock = false;
+                beginEndNestedDepth--;
+                if (beginEndNestedDepth < 0) {
+                    throw new FlywayException("Syntax error in SQL script: unpaired END; statement");
```

Comment:
I wouldn't worry so much about this. A warning is probably enough here. If the script is incorrect (in itself, not due to our parsing) the DB will tell the user.